### DISCUSSION TITLE
Testing and distributing plugins

### DISCUSSION THREAD

**Discussion by laughedelic** [intuit/auto#1963] [Q&A]
Hi! This is about https://github.com/intuit/auto/pull/1962. 

I was wondering what is the recommended way to test a plugin in a non-npm project. Can I publish a canary release from my fork and use the "plugins-included" binary of Auto? Or should I place my plugin in a separate repo and somehow assemble it there?

And after publishing canary versions and testing, what is the proper way to distribute a plugin outside of this monorepo? Do I publish it to NPM? How do I refer to such plugin in the Auto configuration? (non-npm project, so no `npm/yarn install`)

I've seen https://github.com/intuit/auto/discussions/1868, but it was mostly about referring to a local plugin. What about remote plugins? I actually followed the hint from that discussion and used ncc to build a standalone js file with the plugin and all its dependencies, downloaded it to the testing project (in CI) and referred to it as a local plugin. But this seems to be a very cumbersome process. Is there an easier way?

**Comment by hipstersmoothie** [comment]
I just restarted the workflow and you should get a canary version (non-nom) posted to the PR. Thanks for doing the work for this!

> And after publishing canary versions and testing, what is the proper way to distribute a plugin outside of this monorepo? Do I publish it to NPM? How do I refer to such plugin in the Auto configuration? (non-npm project, so no npm/yarn install)

Yeah you should be able to distribute via an npm plugin, install it globally or locally to the project, and use it as normal. You can also write plugins in pure TS without a build step. I also added support to distribute plugins via a shared autorc if you don't want to publish individual plugins.

**Reply by laughedelic** [reply]
> I just restarted the workflow and you should get a canary version (non-nom) posted to the PR. Thanks for doing the work for this!

Thanks! I just tried it out, but something went wrong: https://github.com/intuit/auto/pull/1962#issuecomment-827003423.



> Yeah you should be able to distribute via an npm plugin, install it globally or locally to the project, and use it as normal.

~~So if I publish it to npm, should I use some namespace like `@auto-it/sbt` or is there any naming convention for third-party plugins?~~ (found in the [docs](https://intuit.github.io/auto/docs/configuration/plugins#npm-package))

> You can also write plugins in pure TS without a build step.

I would love to skip the build step. How do I do that? I also saw it mentioned in the [docs](https://intuit.github.io/auto/docs/plugins/writing-plugins), so it says just write this TS/JS file with the import from `auto`, but then I don't understand how to _use_ such plugin. I would still need to publish it, no? Or can I just refer to this single TS file via a direct URL?

> I also added support to **distribute plugins via a shared autorc** if you don't want to publish individual plugins.

Could you elaborate on this a bit too? Is it documented? I double-checked [this doc](https://intuit.github.io/auto/docs/configuration/plugins#npm-package) but don't see it mentioned there.
